src/gridfs/upload.rs

```diff
@@ -0,0 +1,147 @@
+use std::{marker::Unpin, sync::atomic::Ordering};
+
+use futures_util::{
+    io::{AsyncRead, AsyncReadExt},
+    stream::TryStreamExt,
+};
+
+use super::{options::GridFsUploadOptions, Chunk, FilesCollectionDocument, GridFsBucket};
+use crate::{
+    bson::{doc, oid::ObjectId, spec::BinarySubtype, Bson, DateTime, Document, RawBinaryRef},
+    bson_util::get_int,
+    error::{ErrorKind, Result},
+    index::IndexModel,
+    options::{FindOneOptions, ReadPreference, SelectionCriteria},
+    Collection,
+};
+
+impl GridFsBucket {
+    /// Uploads a user file to a GridFS bucket. Bytes are read from `source` and stored in chunks in
+    /// the bucket's chunks collection. After all the chunks have been uploaded, a corresponding
+    /// [`FilesCollectionDocument`] is stored in the bucket's files collection.
+    ///
+    /// This method generates an [`ObjectId`] for the `files_id` field of the
+    /// [`FilesCollectionDocument`] and returns it.
+    pub async fn upload_from_futures_0_3_reader<T>(
+        &self,
+        filename: impl AsRef<str>,
+        source: T,
+        options: impl Into<Option<GridFsUploadOptions>>,
+    ) -> Result<ObjectId>
+    where
+        T: AsyncRead + Unpin,
+    {
+        let id = ObjectId::new();
+        self.upload_from_futures_0_3_reader_with_id(id.into(), filename, source, options)
+            .await?;
+        Ok(id)
+    }
+
+    /// Uploads a user file to a GridFS bucket with the given `files_id`. Bytes are read from
+    /// `source` and stored in chunks in the bucket's chunks collection. After all the chunks have
+    /// been uploaded, a corresponding [`FilesCollectionDocument`] is stored in the bucket's files
+    /// collection.
+    pub async fn upload_from_futures_0_3_reader_with_id<T>(
+        &self,
+        files_id: Bson,
+        filename: impl AsRef<str>,
+        mut source: T,
+        options: impl Into<Option<GridFsUploadOptions>>,
+    ) -> Result<()>
+    where
+        T: AsyncRead + Unpin,
+    {
+        let options = options.into();
+
+        self.create_indexes().await?;
+
+        let chunk_size = options
+            .as_ref()
+            .and_then(|opts| opts.chunk_size_bytes)
+            .unwrap_or_else(|| self.chunk_size_bytes());
+        let mut length = 0u64;
+        let mut n = 0;
+
+        loop {
+            let mut buf = vec![0u8; chunk_size as usize];
+            let bytes_read = match source.read(&mut buf).await {
+                Ok(0) => break,
+                Ok(n) => n,
+                Err(error) => {
```

Comment:
The mechanism to abort an upload is more relevant for `open_download_stream`, when users are incrementally writing their chunks to the stream and closing it out themselves. I think a user could possibly somehow close the source while this method is ongoing, but that would be addressed here, as the call to `read` would fail.

Reply:
ah that makes sense. thanks 👍